subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_df.c

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdint.h>
+#include <errno.h>
+#include <devicetree.h>
+#include <sys/util_macro.h>
+#include <nrfx/hal/nrf_radio.h>
+
+#include "radio_df.h"
+
+/* @brief Minimum antennae number required if antenna switching is enabled */
+#define DF_ANT_NUM_MIN 2
+/* @brief Value to set antenna GPIO pin as not connected */
+#define DF_GPIO_PIN_NOT_SET 0xFF
+/* @brief Number of PSEL_DFEGPIO registers in Radio peripheral */
+#define DF_PSEL_GPIO_NUM 8
+
+/* @brief Direction Finding antenna matrix configuration */
+struct df_ant_cfg {
+	uint8_t ant_num;
+	/* Maximum supported antenna switch pattern length */
+	uint8_t max_ant_sw_pattern_num;
+	/* Selection of GPIOs to be used to switch antennae by Radio */
+	uint8_t dfe_gpio[DF_PSEL_GPIO_NUM];
+};
+
+#define DFE_ANT DT_NODELABEL(dfe_ant)
+#define DFE_GPIO_PIN(idx)						\
+	UTIL_OR(UTIL_AND(DT_NODE_HAS_PROP(DFE_ANT, dfegpio##idx##_pin),	\
+				DT_PROP(DFE_ANT, dfegpio##idx##_pin)),	\
+		DF_GPIO_PIN_NOT_SET)
+
+#define DFE_GPIO_PIN_DISCONNECT (RADIO_PSEL_DFEGPIO_CONNECT_Disconnected << \
+				 RADIO_PSEL_DFEGPIO_CONNECT_Pos)
+
+#if IS_ENABLED(CONFIG_BT_CTLR_DF_ANT_SWITCH_TX) || \
+	IS_ENABLED(CONFIG_BT_CTLR_DF_ANT_SWITCH_RX)
+
+#if DT_NODE_HAS_STATUS(DFE_ANT, okay)
+const static struct df_ant_cfg ant_cfg = {
+	.ant_num = DT_PROP(DFE_ANT, ant_num),
+	.dfe_gpio = {
+		DFE_GPIO_PIN(0),
+		DFE_GPIO_PIN(1),
+		DFE_GPIO_PIN(2),
+		DFE_GPIO_PIN(3),
+		DFE_GPIO_PIN(4),
+		DFE_GPIO_PIN(5),
+		DFE_GPIO_PIN(6),
+		DFE_GPIO_PIN(7),
+	}
+};
+#else
+#error "DF antenna switching feature requires dfe_ant to be enabled in DTS"
+#endif
+
+#endif /* CONFIG_BT_CTLR_DF_ANT_SWITCH_TX ||
+	* CONFIG_BT_CTLR_DF_ANT_SWITCH_RX
+	*/
+
+/* @brief Function performs steps related with DF antennae configuration.
+ *
+ * Validates DTS related with DF antennae.
+ * Sets up DF related PSEL.DFEGPIO registers to give possibility to Radio
+ * to drivee antennae switches.
+ *
+ * @return	Zero in case of success, other value in case of failure.
+ */
+int radio_df_ant_configure(void)
+{
+#if IS_ENABLED(CONFIG_BT_CTLR_DF_ANT_SWITCH_TX) || \
+	IS_ENABLED(CONFIG_BT_CTLR_DF_ANT_SWITCH_RX)
+	uint8_t pins_num = 0;
+
+	if (ant_cfg.ant_num == 0) {
+		return -ENOTSUP;
+	}
+
+	for (uint8_t idx = 0; idx < DF_PSEL_GPIO_NUM; ++idx) {
+		if (ant_cfg.dfe_gpio[idx] != DF_GPIO_PIN_NOT_SET) {
+			++pins_num;
+		}
+	}
+
+	/* Check if there is enough pins configured to represent each pattern
+	 * for given antennae number. Confirm also that antennae number is not
+	 * lower than required minimum.
+	 */
+	if (((0x1 << pins_num) - 1) < ant_cfg.ant_num ||
```

Comment:
Maybe using `BIT_MASK(pins_num)` here would make it a bit clearer.

Reply:
will change

Reply:
code removed due to: https://github.com/zephyrproject-rtos/zephyr/pull/30262/commits/44bfa9ba4f31e5cacddccd1fb2969c30a1dabc6b#r545135430